pyhf/pdf.py

```diff
@@ -181,36 +220,33 @@ def expected_sample(self, channel, sample, pars):
         # \phi == normfactor, overallsys
         # \sigma == histosysdelta + nominal
 
-        tensorlib, _ = get_backend()
         # first, collect the factors from all modifiers
-        results = {'shapesys': [],
-                   'normfactor': [],
-                   'shapefactor': [],
-                   'staterror': [],
-                   'histosys': [],
-                   'normsys': []}
-        for m in sample['modifiers']:
-            modifier, modpars = self.config.modifier(m['name']), pars[self.config.par_slice(m['name'])]
-            results[m['type']].append(modifier.apply(channel, sample, modpars))
-
-
-        # start building the entire set of factors
-        factors = []
-        # scalars that get broadcasted to shape of vectors
-        factors += results['normfactor']
-        factors += results['normsys']
-        # vectors
-        factors += results['shapesys']
-        factors += results['shapefactor']
-        factors += results['staterror']
-
-        nominal = tensorlib.astensor(sample['data'])
-        factors += [tensorlib.sum(tensorlib.stack([
-          nominal,
-          tensorlib.sum(tensorlib.stack(results['histosys']), axis=0)
-        ]), axis=0) if len(results['histosys']) > 0 else nominal]
 
-        return tensorlib.product(tensorlib.stack(tensorlib.simple_broadcast(*factors)), axis=0)
+        all_modifications = {}
+        factor_mods = ['normfactor','normsys','shapesys','shapefactor','staterror']
+        delta_mods  = ['histosys']
+
+        all_results = {}
+        for mtype in factor_mods + delta_mods:
+            all_results[mtype] = self._mtype_results(mtype,pars)
+
+        for channel in self.spec['channels']:
+            for sample in channel['samples']:
+                #sum of lists is concat
+                factors = sum([
```

Comment:
is this supposed to be multiplication and not summation?

Reply:
this is a concatenation of lists .. sum([list1,list3,...],[]) concatenates

Reply:
That's not very clear from the code, or the comment. But ok.